File: src/day1.hs
import Util.String (splitOn)
import Data.List (sort)
import qualified Data.Map as Map

main :: IO ()

getDistance :: ([Int], [Int]) -> [Int]
getDistance (left, right) = zipWith subtract left right

countOccurrences :: (Ord a) => [a] -> Map.Map a Int
countOccurrences = foldr (\x acc -> Map.insertWith (+) x 1 acc) Map.empty

getSimilarity :: [Int] -> Map.Map Int Int -> [Int]
getSimilarity list counts = map (\x -> Map.findWithDefault 0 x counts * x) list

main = do
  contents <- readFile "input/day1.txt"
  let lines = splitOn "\n" contents
  let parts = filter (not . null) (map words lines)
  let numbers = map (map read :: [String] -> [Int]) parts

  let sortLeft = sort (map head numbers)
  let sortRight = sort (map last numbers)

  let distance = getDistance (sortLeft, sortRight)
  let part1 = sum (map abs distance)

  putStrLn $ "Part 1: " ++ show part1

  let counts = countOccurrences sortRight
  let similarities = getSimilarity sortLeft counts

  let part2 = sum similarities

  putStrLn $ "Part 2: " ++ show part2

File: src/day1.ts
import path from 'node:path';
import fs from 'node:fs';

const file = path.join('input/day1.txt');
const content = fs.readFileSync(file, 'utf8');

const lines = content.split('\n').map(l => l.replace(/\s+/g, ' ')).filter(l => l.trim().length > 0);
const numbers = lines.map((line) => line.split(' ').map((num) => parseInt(num, 10)));

const sortLeft = numbers.map(([a]) => a).sort();
const sortRight = numbers.map(([, b]) => b).sort();

const diff = sortLeft.map((a, i) => Math.abs(a - sortRight[i]));
const sum = diff.reduce((a, b) => a + b, 0);

console.log('Part 1:', sum);

const count: Record<number, number> = {};

for (const n of sortRight) {
  count[n] = (count[n] || 0) + 1;
}

let sum2 = 0;
for (const n of sortLeft) {
  sum2 += n * (count[n] || 0);
}

console.log('Part 2:', sum2);
